org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/ProjectManager.xtend

```diff
@@ -105,17 +105,26 @@ class ProjectManager {
     }
 
     protected def XtextResourceSet createFreshResourceSet(ResourceDescriptionsData newIndex) {
-        resourceSetProvider.get => [
-            projectDescription.attachToEmfObject(it)
-            ProjectConfigAdapter.install(it, projectConfig)
-            val index = new ChunkedResourceDescriptions(indexProvider.get, it)
-            index.setContainer(projectDescription.name, newIndex)
-            externalContentSupport.configureResourceSet(it, openedDocumentsContentProvider)
-        ]
+        if (this.resourceSet === null) {
+            this.resourceSet = resourceSetProvider.get => [
+                projectDescription.attachToEmfObject(it)
+                ProjectConfigAdapter.install(it, projectConfig)
+                val index = new ChunkedResourceDescriptions(indexProvider.get, it)
+                index.setContainer(projectDescription.name, newIndex)
+                externalContentSupport.configureResourceSet(it, openedDocumentsContentProvider)
+            ]
+        } else {
+            val resDescs = ChunkedResourceDescriptions.findInEmfObject(this.resourceSet);
+            resDescs.setContainer(projectDescription.name, newIndex)   
+        }
+        return this.resourceSet;
     }
     
     def Resource getResource(URI uri) {
-        resourceSet.getResource(uri, true)
+        val resource = resourceSet.getResource(uri, true)
```

Comment:
Loading it should be ok. We just need to make sure that it is not done in parallel.
I.e. we must not allow parallel read access if we are not sure that everything has been fully initialized.